lib/proxy/client.go

```diff
@@ -0,0 +1,297 @@
+}
+
+type connections struct{}
+
+// NewClient creats a new peer proxy client.
+func NewClient(config ClientConfig) (*Client, error) {
+	err := config.checkAndSetDefaults()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	metrics, err := newClientMetrics()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	c := &Client{
+		done:    make(chan struct{}),
+		config:  config,
+		conns:   make(map[string]*grpc.ClientConn),
+		metrics: metrics,
+	}
+
+	cleanupTicker := config.Clock.NewTicker(config.CleanupInterval)
+	go func() {
+		for {
+			select {
+			case <-c.done:
+				return
+			case <-cleanupTicker.Chan():
+				c.Clean()
+			}
+		}
+	}()
+
+	return c, nil
+}
+
+// Dial dials a node through a peer proxy.
+func (c *Client) Dial(
+	ctx context.Context,
+	proxyAddr string,
+	src net.Addr,
+	dst net.Addr,
+	nodeID string,
+	tunnelType types.TunnelType,
+) (net.Conn, error) {
+	stream, _, err := c.dial(ctx, proxyAddr)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// send dial request as the first frame
+	if err = stream.Send(&clientapi.Frame{
+		Message: &clientapi.Frame_DialRequest{
+			DialRequest: &clientapi.DialRequest{
+				NodeID:     nodeID,
+				TunnelType: tunnelType,
+				Source: &clientapi.NetAddr{
+					Addr:    src.String(),
+					Network: src.Network(),
+				},
+				Destination: &clientapi.NetAddr{
+					Addr:    dst.String(),
+					Network: dst.Network(),
+				},
+			},
+		},
+	}); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	conn := newStreamConn(stream, src, dst)
+	go conn.start()
+
+	return conn, nil
+}
+
+// Close closes all existing client connections.
+func (c *Client) Close() error {
+	c.Lock()
+	defer c.Unlock()
+
+	c.done <- struct{}{}
+
+	var errs []error
+	for _, conn := range c.conns {
+		if err := conn.Close(); err != nil {
+			errs = append(errs, err)
+		}
+	}
+	return trace.NewAggregate(errs...)
+}
+
+// Clean cleans failed client connections.
+func (c *Client) Clean() {
+	c.Lock()
+	defer c.Unlock()
+
+	for k, conn := range c.conns {
+		if failedState(conn.GetState()) {
+			delete(c.conns, k)
+		}
+	}
+}
+
+// dial returns an existing grpc.ClientConn or initializes a new connection to
+// proxyAddr otherwise.
+// The boolean returned in the second argument is intended for testing purposes,
+// to indicates whether the connection was cached or newly established
+func (c *Client) dial(ctx context.Context, proxyAddr string) (clientapi.ProxyService_DialNodeClient, bool, error) {
+	c.Lock()
+	defer c.Unlock()
+
+	stream, err := c.getExistingConnection(ctx, proxyAddr)
+	if err == nil {
+		c.config.Log.Debugf("found existing connection to proxy %+v", proxyAddr)
+		return stream, true, nil
+	} else if !trace.IsNotFound(err) {
+		c.config.Log.Debugf("error checking for existing connections to proxy %+v", proxyAddr)
+		return nil, false, trace.Wrap(err)
+	}
+
+	c.config.Log.Debugf("establishing new connection to proxy %+v. err: %+v", proxyAddr, err)
+
+	stream, err = c.newConnection(ctx, proxyAddr)
```

Comment:
Creating a new connection while we hold the lock could cause some problems.

Also, I think ideally we should try to set up connections before hand rather than as needed. This would avoid slow session starts for the first connection between proxies. But maybe this is a premature optimization? 

The idea would be to use the Cache/Watcher to be able to create/cleanup connections as proxies are added and deleted. This could also help to avoid removing a client that happens to be in a temporary `TransientFailure` state during cleanup.

Reply:
> Also, I think ideally we should try to set up connections before hand rather than as needed. This would avoid slow session starts for the first connection between proxies. But maybe this is a premature optimization?

No it's all good. I thought we weren't going to include the watcher in this library, and have him call this `dial` function after it got exported. But it you know what now that you mention it, it will look cleaner. I'll work on it.